app/crypto/digests.c

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright (c) 2015, DiUS Computing Pty Ltd (jmattsson@dius.com.au)
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the copyright holder nor the names of contributors
+ *    may be used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTOR(S) ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTOR(S) BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ *
+ */
+#include "digests.h"
+#include "lwip/mem.h"
+#include "lwip/arch.h"
+#include "ssl/ssl_crypto.h"
+#include "sha2.h"
+#include <string.h>
+#include <c_errno.h>
+
+typedef char ensure_int_and_size_t_same[(sizeof(int)==sizeof(size_t)) ? 0 : -1];
+
+/* None of the functions match the prototype fully due to the void *, and in
+   some cases also the int vs size_t len, so wrap declarations in a macro. */
+#define MECH(pfx, ds, bs) \
+  { #pfx, \
+    (create_ctx_fn)pfx ## _Init, \
+    (update_ctx_fn)pfx ## _Update, \
+    (finalize_ctx_fn)pfx ## _Final, \
+    sizeof(pfx ## _CTX), \
+    ds, \
+    bs }
+
+static const digest_mech_info_t hash_mechs[] =
+{
+   MECH(MD2,  MD2_SIZE,  16)
+  ,MECH(MD5,  MD5_SIZE,  64)
+  ,MECH(SHA1, SHA1_SIZE, 64)
+#ifndef WITHOUT_SHA2
+  ,MECH(SHA256, SHA256_DIGEST_LENGTH, SHA256_BLOCK_LENGTH)
+  ,MECH(SHA384, SHA384_DIGEST_LENGTH, SHA384_BLOCK_LENGTH)
+  ,MECH(SHA512, SHA512_DIGEST_LENGTH, SHA512_BLOCK_LENGTH)
+#endif
+};
+
+#undef MECH
+
+const digest_mech_info_t *crypto_digest_mech (const char *mech)
+{
+  if (!mech)
+    return 0;
+
+  size_t i;
+  for (i = 0; i < (sizeof (hash_mechs) / sizeof (digest_mech_info_t)); ++i)
+  {
+    const digest_mech_info_t *mi = hash_mechs + i;
+    if (strcasecmp (mech, mi->name) == 0)
+      return mi;
+  }
+  return 0;
+}
+
+static const char hex[] = "0123456789abcdef";
```

Comment:
FWIW, I already had a hex encoder in my crypto PR (though as a lua function and not internal C function like this one).

Is this used internally by other C functions?